Realtime test sometime failing?

The test `tests/test_realtime.py::test_realtime_cache_different_settings[duckdb]` failed in python 3.9 [on this run](https://github.com/ADBond/splink/actions/runs/11932686224/job/33258155342?pr=41). Re-running the job allowed the test to pass, which is odd as I can't see where any indeterminacy could occur.

`Error was: Binder Error: Values list "l" does not have a column named "tf_first_name"`.

<details>
<summary>Full test output</summary>

```
________________ test_realtime_cache_different_settings[duckdb] ________________

self = <splink.internals.duckdb.database_api.DuckDBAPI object at 0x7fb7780ef4c0>
final_sql = 'CREATE TABLE __splink__realtime_compare_records_aj2njrcn AS\nWITH __splink__compare_records_left AS (\n  SELECT\n    ..._adj_city,\n  "email_l",\n  "email_r",\n  gamma_email,\n  bf_email,\n  match_key\nFROM __splink__df_match_weight_parts'
templated_name = '__splink__realtime_compare_records'
physical_name = '__splink__realtime_compare_records_aj2njrcn'

    @final
    def _log_and_run_sql_execution(
        self, final_sql: str, templated_name: str, physical_name: str
    ) -> TablishType:
        """
        Log some sql, then call _run_sql_execution()
        Any errors will be converted to SplinkException with more detail
        names are only relevant for logging, not execution
        """
        logger.debug(execute_sql_logging_message_info(templated_name, physical_name))
        logger.log(5, log_sql(final_sql))
        try:
>           return self._execute_sql_against_backend(final_sql)

splink/internals/database_api.py:63: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <splink.internals.duckdb.database_api.DuckDBAPI object at 0x7fb7780ef4c0>
final_sql = 'CREATE TABLE __splink__realtime_compare_records_aj2njrcn AS \nWITH\n\n__splink__compare_records_left as (\nselect * f..._tf_adj_city,"email_l","email_r",gamma_email,bf_email,match_key \n    from __splink__df_match_weight_parts\n    \n    '

    def _execute_sql_against_backend(self, final_sql: str) -> duckdb.DuckDBPyRelation:
>       return self._con.sql(final_sql)
E       duckdb.duckdb.BinderException: Binder Error: Values list "l" does not have a column named "tf_first_name"
E       LINE 11: ...from __splink__compare_records_left_aj2njrcn), 
E       
E       __splink__compare_records_rig...
E                                                          ^

splink/internals/duckdb/database_api.py:102: BinderException

The above exception was the direct cause of the following exception:

test_helpers = {'duckdb': (<class 'tests.helpers.DuckDBTestHelper'>, []), 'spark': (<class 'tests.helpers.SparkTestHelper'>, [<functi...ite': (<class 'tests.helpers.SQLiteTestHelper'>, []), 'postgres': (<class 'tests.helpers.PostgresTestHelper'>, [None])}
dialect = 'duckdb'

    @mark_with_dialects_excluding()
    def test_realtime_cache_different_settings(test_helpers, dialect):
        helper = test_helpers[dialect]
        db_api = helper.extra_linker_args()["db_api"]
    
        df1 = pd.DataFrame(
            [
                {
                    "unique_id": 0,
                    "first_name": "Julia",
                    "surname": "Taylor",
                    "city": "London",
                    "email": "julia@email.com",
                }
            ]
        )
    
        df2 = pd.DataFrame(
            [
                {
                    "unique_id": 1,
                    "first_name": "Julia",
                    "surname": "Taylor",
                    "city": "London",
                    "email": "bad@address.com",
                }
            ]
        )
    
        settings_1 = SettingsCreator(
            link_type="dedupe_only",
            comparisons=[
                cl.ExactMatch("first_name"),
                cl.ExactMatch("surname"),
                cl.ExactMatch("city"),
            ],
            blocking_rules_to_generate_predictions=[block_on("first_name")],
        )
    
        settings_2 = SettingsCreator(
            link_type="dedupe_only",
            comparisons=[
                cl.ExactMatch("first_name"),
                cl.ExactMatch("surname"),
                cl.ExactMatch("email"),
            ],
            blocking_rules_to_generate_predictions=[block_on("first_name")],
        )
    
>       res1 = compare_records(
            df1, df2, settings_1, db_api, use_sql_from_cache=True
        ).as_record_dict()[0]["match_weight"]

tests/test_realtime.py:343: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
splink/internals/realtime.py:88: in compare_records
    return db_api._sql_to_splink_dataframe(
splink/internals/database_api.py:93: in _sql_to_splink_dataframe
    spark_df = self._log_and_run_sql_execution(sql, templated_name, physical_name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <splink.internals.duckdb.database_api.DuckDBAPI object at 0x7fb7780ef4c0>
final_sql = 'CREATE TABLE __splink__realtime_compare_records_aj2njrcn AS\nWITH __splink__compare_records_left AS (\n  SELECT\n    ..._adj_city,\n  "email_l",\n  "email_r",\n  gamma_email,\n  bf_email,\n  match_key\nFROM __splink__df_match_weight_parts'
templated_name = '__splink__realtime_compare_records'
physical_name = '__splink__realtime_compare_records_aj2njrcn'

    @final
    def _log_and_run_sql_execution(
        self, final_sql: str, templated_name: str, physical_name: str
    ) -> TablishType:
        """
        Log some sql, then call _run_sql_execution()
        Any errors will be converted to SplinkException with more detail
        names are only relevant for logging, not execution
        """
        logger.debug(execute_sql_logging_message_info(templated_name, physical_name))
        logger.log(5, log_sql(final_sql))
        try:
            return self._execute_sql_against_backend(final_sql)
        except Exception as e:
            # Parse our SQL through sqlglot to pretty print
            try:
                final_sql = sqlglot.parse_one(
                    final_sql,
                    read=self.sql_dialect.sqlglot_dialect,
                ).sql(pretty=True)
                # if sqlglot produces any errors, just report the raw SQL
            except Exception:
                pass
    
>           raise SplinkException(
                f"Error executing the following sql for table "
                f"`{templated_name}`({physical_name}):\n{final_sql}"
                f"\n\nError was: {e}"
            ) from e
E           splink.internals.exceptions.SplinkException: Error executing the following sql for table `__splink__realtime_compare_records`(__splink__realtime_compare_records_aj2njrcn):
E           CREATE TABLE __splink__realtime_compare_records_aj2njrcn AS
E           WITH __splink__compare_records_left AS (
E             SELECT
E               *
E             FROM __splink__compare_records_left_aj2njrcn
E           ), __splink__compare_records_right AS (
E             SELECT
E               *
E             FROM __splink__compare_records_right_aj2njrcn
E           ), __splink__compare_two_records_blocked AS (
E             SELECT
E               "l"."unique_id" AS "unique_id_l",
E               "r"."unique_id" AS "unique_id_r",
E               "l"."first_name" AS "first_name_l",
E               "r"."first_name" AS "first_name_r",
E               "l"."tf_first_name" AS "tf_first_name_l",
E               "r"."tf_first_name" AS "tf_first_name_r",
E               "l"."surname" AS "surname_l",
E               "r"."surname" AS "surname_r",
E               "l"."city" AS "city_l",
E               "r"."city" AS "city_r",
E               "l"."tf_city" AS "tf_city_l",
E               "r"."tf_city" AS "tf_city_r",
E               "l"."email" AS "email_l",
E               "r"."email" AS "email_r",
E               0 AS match_key
E             FROM __splink__compare_records_left AS l
E             CROSS JOIN __splink__compare_records_right AS r
E           ), __splink__df_comparison_vectors AS (
E             SELECT
E               "unique_id_l",
E               "unique_id_r",
E               "first_name_l",
E               "first_name_r",
E               CASE
E                 WHEN "first_name_l" IS NULL OR "first_name_r" IS NULL
E                 THEN -1
E                 WHEN "first_name_l" = "first_name_r"
E                 THEN 1
E                 ELSE 0
E               END AS gamma_first_name,
E               "tf_first_name_l",
E               "tf_first_name_r",
E               "surname_l",
E               "surname_r",
E               CASE
E                 WHEN "surname_l" IS NULL OR "surname_r" IS NULL
E                 THEN -1
E                 WHEN "surname_l" = "surname_r"
E                 THEN 1
E                 ELSE 0
E               END AS gamma_surname,
E               "city_l",
E               "city_r",
E               CASE
E                 WHEN "city_l" IS NULL OR "city_r" IS NULL
E                 THEN -1
E                 WHEN "city_l" = "city_r"
E                 THEN 1
E                 ELSE 0
E               END AS gamma_city,
E               "tf_city_l",
E               "tf_city_r",
E               "email_l",
E               "email_r",
E               CASE
E                 WHEN "email_l" IS NULL OR "email_r" IS NULL
E                 THEN -1
E                 WHEN "email_l" = "email_r"
E                 THEN 1
E                 ELSE 0
E               END AS gamma_email,
E               match_key
E             FROM __splink__compare_two_records_blocked
E           ), __splink__df_match_weight_parts AS (
E             SELECT
E               "unique_id_l",
E               "unique_id_r",
E               "first_name_l",
E               "first_name_r",
E               gamma_first_name,
E               "tf_first_name_l",
E               "tf_first_name_r",
E               CASE
E                 WHEN gamma_first_name = -1
E                 THEN CAST(1.0 AS DOUBLE)
E                 WHEN gamma_first_name = 1
E                 THEN CAST(1024.0 AS DOUBLE)
E                 WHEN gamma_first_name = 0
E                 THEN CAST(0.03125 AS DOUBLE)
E               END AS bf_first_name,
E               CASE
E                 WHEN gamma_first_name = -1
E                 THEN CAST(1 AS DOUBLE)
E                 WHEN gamma_first_name = 1
E                 THEN (
E                   CASE
E                     WHEN NOT COALESCE("tf_first_name_l", "tf_first_name_r") IS NULL
E                     THEN POWER(
E                       CAST(0.000927734375 AS DOUBLE) / NULLIF(
E                         (
E                           CASE
E                             WHEN COALESCE("tf_first_name_l", "tf_first_name_r") >= COALESCE("tf_first_name_r", "tf_first_name_l")
E                             THEN COALESCE("tf_first_name_l", "tf_first_name_r")
E                             ELSE COALESCE("tf_first_name_r", "tf_first_name_l")
E                           END
E                         ),
E                         0
E                       ),
E                       CAST(1.0 AS DOUBLE)
E                     )
E                     ELSE CAST(1 AS DOUBLE)
E                   END
E                 )
E                 WHEN gamma_first_name = 0
E                 THEN CAST(1 AS DOUBLE)
E               END AS bf_tf_adj_first_name,
E               "surname_l",
E               "surname_r",
E               gamma_surname,
E               CASE
E                 WHEN gamma_surname = -1
E                 THEN CAST(1.0 AS DOUBLE)
E                 WHEN gamma_surname = 1
E                 THEN CAST(1024.0 AS DOUBLE)
E                 WHEN gamma_surname = 0
E                 THEN CAST(0.03125 AS DOUBLE)
E               END AS bf_surname,
E               "city_l",
E               "city_r",
E               gamma_city,
E               "tf_city_l",
E               "tf_city_r",
E               CASE
E                 WHEN gamma_city = -1
E                 THEN CAST(1.0 AS DOUBLE)
E                 WHEN gamma_city = 1
E                 THEN CAST(1024.0 AS DOUBLE)
E                 WHEN gamma_city = 0
E                 THEN CAST(0.03125 AS DOUBLE)
E               END AS bf_city,
E               CASE
E                 WHEN gamma_city = -1
E                 THEN CAST(1 AS DOUBLE)
E                 WHEN gamma_city = 1
E                 THEN (
E                   CASE
E                     WHEN NOT COALESCE("tf_city_l", "tf_city_r") IS NULL
E                     THEN POWER(
E                       CAST(0.000927734375 AS DOUBLE) / NULLIF(
E                         (
E                           CASE
E                             WHEN COALESCE("tf_city_l", "tf_city_r") >= COALESCE("tf_city_r", "tf_city_l")
E                             THEN COALESCE("tf_city_l", "tf_city_r")
E                             ELSE COALESCE("tf_city_r", "tf_city_l")
E                           END
E                         ),
E                         0
E                       ),
E                       CAST(1.0 AS DOUBLE)
E                     )
E                     ELSE CAST(1 AS DOUBLE)
E                   END
E                 )
E                 WHEN gamma_city = 0
E                 THEN CAST(1 AS DOUBLE)
E               END AS bf_tf_adj_city,
E               "email_l",
E               "email_r",
E               gamma_email,
E               CASE
E                 WHEN gamma_email = -1
E                 THEN CAST(1.0 AS DOUBLE)
E                 WHEN gamma_email = 1
E                 THEN CAST(1024.0 AS DOUBLE)
E                 WHEN gamma_email = 0
E                 THEN CAST(0.03125 AS DOUBLE)
E               END AS bf_email,
E               match_key
E             FROM __splink__df_comparison_vectors
E           )
E           SELECT
E             LOG(
E               2,
E               CAST(0.00010001000100010001 AS DOUBLE) * bf_first_name * bf_tf_adj_first_name * bf_surname * bf_city * bf_tf_adj_city * bf_email
E             ) AS match_weight,
E             CASE
E               WHEN bf_first_name = CAST('infinity' AS DOUBLE)
E               OR bf_tf_adj_first_name = CAST('infinity' AS DOUBLE)
E               OR bf_surname = CAST('infinity' AS DOUBLE)
E               OR bf_city = CAST('infinity' AS DOUBLE)
E               OR bf_tf_adj_city = CAST('infinity' AS DOUBLE)
E               OR bf_email = CAST('infinity' AS DOUBLE)
E               THEN 1.0
E               ELSE (
E                 CAST(0.00010001000100010001 AS DOUBLE) * bf_first_name * bf_tf_adj_first_name * bf_surname * bf_city * bf_tf_adj_city * bf_email
E               ) / NULLIF(
E                 (
E                   1 + (
E                     CAST(0.00010001000100010001 AS DOUBLE) * bf_first_name * bf_tf_adj_first_name * bf_surname * bf_city * bf_tf_adj_city * bf_email
E                   )
E                 ),
E                 0
E               )
E             END AS match_probability,
E             "unique_id_l",
E             "unique_id_r",
E             "first_name_l",
E             "first_name_r",
E             gamma_first_name,
E             "tf_first_name_l",
E             "tf_first_name_r",
E             bf_first_name,
E             bf_tf_adj_first_name,
E             "surname_l",
E             "surname_r",
E             gamma_surname,
E             bf_surname,
E             "city_l",
E             "city_r",
E             gamma_city,
E             "tf_city_l",
E             "tf_city_r",
E             bf_city,
E             bf_tf_adj_city,
E             "email_l",
E             "email_r",
E             gamma_email,
E             bf_email,
E             match_key
E           FROM __splink__df_match_weight_parts
E           
E           Error was: Binder Error: Values list "l" does not have a column named "tf_first_name"
E           LINE 11: ...from __splink__compare_records_left_aj2njrcn), 
E           
E           __splink__compare_records_rig...
E                                                              ^

splink/internals/database_api.py:75: SplinkException
```

</details>

<details>
<summary>Test summary output</summary>

```
=========================== short test summary info ============================
FAILED tests/test_realtime.py::test_realtime_cache_different_settings[duckdb] - splink.internals.exceptions.SplinkException: Error executing the following sql for table `__splink__realtime_compare_records`(__splink__realtime_compare_records_aj2njrcn):
CREATE TABLE __splink__realtime_compare_records_aj2njrcn AS
WITH __splink__compare_records_left AS (
  SELECT
    *
  FROM __splink__compare_records_left_aj2njrcn
), __splink__compare_records_right AS (
  SELECT
    *
  FROM __splink__compare_records_right_aj2njrcn
), __splink__compare_two_records_blocked AS (
  SELECT
    "l"."unique_id" AS "unique_id_l",
    "r"."unique_id" AS "unique_id_r",
    "l"."first_name" AS "first_name_l",
    "r"."first_name" AS "first_name_r",
    "l"."tf_first_name" AS "tf_first_name_l",
    "r"."tf_first_name" AS "tf_first_name_r",
    "l"."surname" AS "surname_l",
    "r"."surname" AS "surname_r",
    "l"."city" AS "city_l",
    "r"."city" AS "city_r",
    "l"."tf_city" AS "tf_city_l",
    "r"."tf_city" AS "tf_city_r",
    "l"."email" AS "email_l",
    "r"."email" AS "email_r",
    0 AS match_key
  FROM __splink__compare_records_left AS l
  CROSS JOIN __splink__compare_records_right AS r
), __splink__df_comparison_vectors AS (
  SELECT
    "unique_id_l",
    "unique_id_r",
    "first_name_l",
    "first_name_r",
    CASE
      WHEN "first_name_l" IS NULL OR "first_name_r" IS NULL
      THEN -1
      WHEN "first_name_l" = "first_name_r"
      THEN 1
      ELSE 0
    END AS gamma_first_name,
    "tf_first_name_l",
    "tf_first_name_r",
    "surname_l",
    "surname_r",
    CASE
      WHEN "surname_l" IS NULL OR "surname_r" IS NULL
      THEN -1
      WHEN "surname_l" = "surname_r"
      THEN 1
      ELSE 0
    END AS gamma_surname,
    "city_l",
    "city_r",
    CASE
      WHEN "city_l" IS NULL OR "city_r" IS NULL
      THEN -1
      WHEN "city_l" = "city_r"
      THEN 1
      ELSE 0
    END AS gamma_city,
    "tf_city_l",
    "tf_city_r",
    "email_l",
    "email_r",
    CASE
      WHEN "email_l" IS NULL OR "email_r" IS NULL
      THEN -1
      WHEN "email_l" = "email_r"
      THEN 1
      ELSE 0
    END AS gamma_email,
    match_key
  FROM __splink__compare_two_records_blocked
), __splink__df_match_weight_parts AS (
  SELECT
    "unique_id_l",
    "unique_id_r",
    "first_name_l",
    "first_name_r",
    gamma_first_name,
    "tf_first_name_l",
    "tf_first_name_r",
    CASE
      WHEN gamma_first_name = -1
      THEN CAST(1.0 AS DOUBLE)
      WHEN gamma_first_name = 1
      THEN CAST(1024.0 AS DOUBLE)
      WHEN gamma_first_name = 0
      THEN CAST(0.03125 AS DOUBLE)
    END AS bf_first_name,
    CASE
      WHEN gamma_first_name = -1
      THEN CAST(1 AS DOUBLE)
      WHEN gamma_first_name = 1
      THEN (
        CASE
          WHEN NOT COALESCE("tf_first_name_l", "tf_first_name_r") IS NULL
          THEN POWER(
            CAST(0.000927734375 AS DOUBLE) / NULLIF(
              (
                CASE
                  WHEN COALESCE("tf_first_name_l", "tf_first_name_r") >= COALESCE("tf_first_name_r", "tf_first_name_l")
                  THEN COALESCE("tf_first_name_l", "tf_first_name_r")
                  ELSE COALESCE("tf_first_name_r", "tf_first_name_l")
                END
              ),
              0
            ),
            CAST(1.0 AS DOUBLE)
          )
          ELSE CAST(1 AS DOUBLE)
        END
      )
      WHEN gamma_first_name = 0
      THEN CAST(1 AS DOUBLE)
    END AS bf_tf_adj_first_name,
    "surname_l",
    "surname_r",
    gamma_surname,
    CASE
      WHEN gamma_surname = -1
      THEN CAST(1.0 AS DOUBLE)
      WHEN gamma_surname = 1
      THEN CAST(1024.0 AS DOUBLE)
      WHEN gamma_surname = 0
      THEN CAST(0.03125 AS DOUBLE)
    END AS bf_surname,
    "city_l",
    "city_r",
    gamma_city,
    "tf_city_l",
    "tf_city_r",
    CASE
      WHEN gamma_city = -1
      THEN CAST(1.0 AS DOUBLE)
      WHEN gamma_city = 1
      THEN CAST(1024.0 AS DOUBLE)
      WHEN gamma_city = 0
      THEN CAST(0.03125 AS DOUBLE)
    END AS bf_city,
    CASE
      WHEN gamma_city = -1
      THEN CAST(1 AS DOUBLE)
      WHEN gamma_city = 1
      THEN (
        CASE
          WHEN NOT COALESCE("tf_city_l", "tf_city_r") IS NULL
          THEN POWER(
            CAST(0.000927734375 AS DOUBLE) / NULLIF(
              (
                CASE
                  WHEN COALESCE("tf_city_l", "tf_city_r") >= COALESCE("tf_city_r", "tf_city_l")
                  THEN COALESCE("tf_city_l", "tf_city_r")
                  ELSE COALESCE("tf_city_r", "tf_city_l")
                END
              ),
              0
            ),
            CAST(1.0 AS DOUBLE)
          )
          ELSE CAST(1 AS DOUBLE)
        END
      )
      WHEN gamma_city = 0
      THEN CAST(1 AS DOUBLE)
    END AS bf_tf_adj_city,
    "email_l",
    "email_r",
    gamma_email,
    CASE
      WHEN gamma_email = -1
      THEN CAST(1.0 AS DOUBLE)
      WHEN gamma_email = 1
      THEN CAST(1024.0 AS DOUBLE)
      WHEN gamma_email = 0
      THEN CAST(0.03125 AS DOUBLE)
    END AS bf_email,
    match_key
  FROM __splink__df_comparison_vectors
)
SELECT
  LOG(
    2,
    CAST(0.00010001000100010001 AS DOUBLE) * bf_first_name * bf_tf_adj_first_name * bf_surname * bf_city * bf_tf_adj_city * bf_email
  ) AS match_weight,
  CASE
    WHEN bf_first_name = CAST('infinity' AS DOUBLE)
    OR bf_tf_adj_first_name = CAST('infinity' AS DOUBLE)
    OR bf_surname = CAST('infinity' AS DOUBLE)
    OR bf_city = CAST('infinity' AS DOUBLE)
    OR bf_tf_adj_city = CAST('infinity' AS DOUBLE)
    OR bf_email = CAST('infinity' AS DOUBLE)
    THEN 1.0
    ELSE (
      CAST(0.00010001000100010001 AS DOUBLE) * bf_first_name * bf_tf_adj_first_name * bf_surname * bf_city * bf_tf_adj_city * bf_email
    ) / NULLIF(
      (
        1 + (
          CAST(0.00010001000100010001 AS DOUBLE) * bf_first_name * bf_tf_adj_first_name * bf_surname * bf_city * bf_tf_adj_city * bf_email
        )
      ),
      0
    )
  END AS match_probability,
  "unique_id_l",
  "unique_id_r",
  "first_name_l",
  "first_name_r",
  gamma_first_name,
  "tf_first_name_l",
  "tf_first_name_r",
  bf_first_name,
  bf_tf_adj_first_name,
  "surname_l",
  "surname_r",
  gamma_surname,
  bf_surname,
  "city_l",
  "city_r",
  gamma_city,
  "tf_city_l",
  "tf_city_r",
  bf_city,
  bf_tf_adj_city,
  "email_l",
  "email_r",
  gamma_email,
  bf_email,
  match_key
FROM __splink__df_match_weight_parts

Error was: Binder Error: Values list "l" does not have a column named "tf_first_name"
LINE 11: ...from __splink__compare_records_left_aj2njrcn), 

__splink__compare_records_rig...
                                                   ^
= 1 failed, 321 passed, 12 skipped, 395 deselected, 16 warnings in 116.40s (0:01:56) =
```

</details>

<details>
<summary>Environment info (3.9.20)</summary>

```
Package operations: 36 installs, 0 updates, 1 removal

  - Removing setuptools (75.5.0)
  - Installing six (1.16.0)
  - Installing attrs (24.2.0)
  - Installing python-dateutil (2.9.0.post0)
  - Installing rpds-py (0.20.1)
  - Installing referencing (0.35.1)
  - Installing jsonschema-specifications (2023.12.1)
  - Installing markupsafe (2.1.5)
  - Installing numpy (1.24.4)
  - Installing pytz (2024.2)
  - Installing tzdata (2024.2)
  - Installing exceptiongroup (1.2.2)
  - Installing greenlet (3.1.1)
  - Installing iniconfig (2.0.0)
  - Installing jinja2 (3.1.4)
  - Installing jsonschema (4.23.0)
  - Installing packaging (24.2)
  - Installing narwhals (1.14.1)
  - Installing pandas (2.2.3)
  - Installing pluggy (1.5.0)
  - Installing py4j (0.10.9.7)
  - Installing pyarrow (17.0.0)
  - Installing texttable (1.7.0)
  - Installing tomli (2.1.0)
  - Installing typing-extensions (4.12.2)
  - Installing altair (5.4.1)
  - Installing duckdb (1.1.3)
  - Installing igraph (0.11.8)
  - Installing networkx (3.1)
  - Installing psycopg2-binary (2.9.10)
  - Installing pyspark (3.5.3)
  - Installing pytest (8.3.3)
  - Installing rapidfuzz (3.9.7)
  - Installing ruff (0.4.10)
  - Installing sqlalchemy (2.0.36)
  - Installing sqlglot (25.31.4)
  - Installing tabulate (0.9.0)
```
</details>

--- a/splink/internals/comparison.py
+++ b/splink/internals/comparison.py
@@ -61,12 +61,18 @@ class Comparison:
 
     def __init__(
         self,
-        comparison_levels: List[ComparisonLevel],
+        comparison_levels: List[ComparisonLevel | dict],
         sqlglot_dialect: str,
         output_column_name: str = None,
         comparison_description: str = None,
         column_info_settings: ColumnInfoSettings = None,
     ):
+        comparison_levels: list[ComparisonLevel] = [
+            ComparisonLevel(**cl, sqlglot_dialect=sqlglot_dialect)
+            if isinstance(cl, dict)
+            else cl
+            for cl in comparison_levels
+        ]
         self.comparison_levels: list[ComparisonLevel] = comparison_levels
 
         self._column_info_settings: Optional[ColumnInfoSettings] = column_info_settings

--- a/splink/internals/comparison.py
+++ b/splink/internals/comparison.py
@@ -61,19 +61,19 @@ class Comparison:
 
     def __init__(
         self,
-        comparison_levels: List[ComparisonLevel | dict],
+        comparison_levels: List[ComparisonLevel | dict[str, Any]],
         sqlglot_dialect: str,
         output_column_name: str = None,
         comparison_description: str = None,
         column_info_settings: ColumnInfoSettings = None,
     ):
-        comparison_levels: list[ComparisonLevel] = [
+        comparison_levels_as_objs: list[ComparisonLevel] = [
             ComparisonLevel(**cl, sqlglot_dialect=sqlglot_dialect)
             if isinstance(cl, dict)
             else cl
             for cl in comparison_levels
         ]
-        self.comparison_levels: list[ComparisonLevel] = comparison_levels
+        self.comparison_levels: list[ComparisonLevel] = comparison_levels_as_objs
 
         self._column_info_settings: Optional[ColumnInfoSettings] = column_info_settings
 

--- a/splink/internals/comparison_creator.py
+++ b/splink/internals/comparison_creator.py
@@ -140,7 +140,7 @@ class ComparisonCreator(ABC):
             "comparison_description": self.create_description(),
             "output_column_name": self.create_output_column_name(),
             "comparison_levels": [
-                cl.get_comparison_level(sql_dialect_str)
+                cl.get_comparison_level(sql_dialect_str).as_dict()
                 for cl in self.get_configured_comparison_levels()
             ],
         }

--- a/splink/internals/comparison_level.py
+++ b/splink/internals/comparison_level.py
@@ -658,6 +658,9 @@ class ComparisonLevel:
         if self._u_probability and self._u_is_trained:
             output["u_probability"] = self.u_probability
 
+        output["fix_m_probability"] = self._fix_m_probability
+        output["fix_u_probability"] = self._fix_u_probability
+
         if self._has_tf_adjustments:
             output["tf_adjustment_column"] = self._tf_adjustment_input_column.input_name
             if self._tf_minimum_u_value != 0:

--- a/splink/internals/realtime.py
+++ b/splink/internals/realtime.py
@@ -15,17 +15,17 @@ from splink.internals.settings_creator import SettingsCreator
 from splink.internals.splink_dataframe import SplinkDataFrame
 
 
+
 class SQLCache:
     def __init__(self):
         self._cache = {}
 
     # TODO: if we have path/string, do we want to think about behaviour if underlying
     # file changes between calls?
-    # TODO: might need to think about how equality works with WeakKeyDictionary in case
-    # there are any gotchas
-    # TODO: maybe check if string interning affects this in some way
-    def get(self, settings: SettingsCreator | dict[str, Any] | Path | str, new_uid: str) -> str | None:
-        settings_id = id(settings)
+    def get(
+        self, settings: SettingsCreator | dict[str, Any] | Path | str, new_uid: str
+    ) -> str | None:
+        settings_id = self._cache_id(settings)
         if settings_id not in self._cache:
             return None
         sql, cached_uid, settings_ref = self._cache[settings_id]
@@ -39,10 +39,33 @@ class SQLCache:
             sql = sql.replace(cached_uid, new_uid)
         return sql
 
-    def set(self, settings: SettingsCreator | dict[str, Any] | Path | str, sql: str | None, uid: str | None) -> None:
-        settings_id = id(settings)
+    def set(
+        self,
+        settings: SettingsCreator | dict[str, Any] | Path | str,
+        sql: str | None,
+        uid: str | None,
+    ) -> None:
         if sql is not None:
-            self._cache[settings_id] = (sql, uid, ref(settings))
+            settings_id = self._cache_id(settings)
+            # kind of hacky
+            # allows us to not need to special-case retrieval - will appear as though
+            # weakref is always live, so don't need to intervene
+            settings_ref = (
+                ref(settings)
+                if isinstance(settings, SettingsCreator)
+                else (lambda x: True)
+            )
+            self._cache[settings_id] = (sql, uid, settings_ref)
+
+    @staticmethod
+    def _cache_id(settings: SettingsCreator | dict[str, Any] | Path | str):
+        if isinstance(settings, SettingsCreator):
+            return str(id(settings))
+        if isinstance(settings, str):
+            return settings
+        if isinstance(settings, Path):
+            return str(settings)
+        # TODO: dict?
 
 
 _sql_cache = SQLCache()

--- a/splink/internals/realtime.py
+++ b/splink/internals/realtime.py
@@ -65,7 +65,7 @@ class SQLCache:
     @staticmethod
     def _cache_id(
         settings: SettingsCreator | dict[str, Any] | Path | str, sql_dialect_str: str
-    ):
+    ) -> str:
         if isinstance(settings, SettingsCreator):
             return str(id(settings))
         if isinstance(settings, str):

--- a/splink/internals/realtime.py
+++ b/splink/internals/realtime.py
@@ -1,5 +1,7 @@
 from __future__ import annotations
 
+import json
+import logging
 from pathlib import Path
 from typing import Any
 from weakref import ref
@@ -23,9 +25,13 @@ class SQLCache:
     # TODO: if we have path/string, do we want to think about behaviour if underlying
     # file changes between calls?
     def get(
-        self, settings: SettingsCreator | dict[str, Any] | Path | str, new_uid: str
+        self,
+        settings: SettingsCreator | dict[str, Any] | Path | str,
+        new_uid: str,
+        *,
+        sql_dialect_str: str,
     ) -> str | None:
-        settings_id = self._cache_id(settings)
+        settings_id = self._cache_id(settings, sql_dialect_str)
         if settings_id not in self._cache:
             return None
         sql, cached_uid, settings_ref = self._cache[settings_id]
@@ -44,28 +50,40 @@ class SQLCache:
         settings: SettingsCreator | dict[str, Any] | Path | str,
         sql: str | None,
         uid: str | None,
+        *,
+        sql_dialect_str: str,
     ) -> None:
         if sql is not None:
-            settings_id = self._cache_id(settings)
+            settings_id = self._cache_id(settings, sql_dialect_str)
             # kind of hacky
             # allows us to not need to special-case retrieval - will appear as though
             # weakref is always live, so don't need to intervene
             settings_ref = (
                 ref(settings)
                 if isinstance(settings, SettingsCreator)
-                else (lambda x: True)
+                else (lambda: True)
             )
             self._cache[settings_id] = (sql, uid, settings_ref)
 
     @staticmethod
-    def _cache_id(settings: SettingsCreator | dict[str, Any] | Path | str):
+    def _cache_id(
+        settings: SettingsCreator | dict[str, Any] | Path | str, sql_dialect_str: str
+    ):
         if isinstance(settings, SettingsCreator):
             return str(id(settings))
         if isinstance(settings, str):
             return settings
         if isinstance(settings, Path):
             return str(settings)
-        # TODO: dict?
+        # we have a dict
+        try:
+            key = json.dumps(settings)
+        except TypeError:
+            settings_dict = SettingsCreator(**settings).create_settings_dict(
+                sql_dialect_str=sql_dialect_str
+            )
+            key = json.dumps(settings_dict)
+        return key
 
 
 _sql_cache = SQLCache()
@@ -93,6 +111,7 @@ def compare_records(
     global _sql_cache
 
     uid = ascii_uid(8)
+    sql_dialect_str = db_api.sql_dialect.sql_dialect_str
 
     if isinstance(record_1, dict):
         to_register_left: AcceptableInputTableType = [record_1]
@@ -118,9 +137,11 @@ def compare_records(
     )
     df_records_right.templated_name = "__splink__compare_records_right"
 
-    settings_id = id(settings)
     if use_sql_from_cache:
-        if cached_sql := _sql_cache.get(settings, uid):
+        cached_sql = _sql_cache.get(
+            settings, uid, sql_dialect_str=sql_dialect_str
+        )
+        if cached_sql:
             return db_api._sql_to_splink_dataframe(
                 cached_sql,
                 templated_name="__splink__realtime_compare_records",
@@ -173,6 +194,8 @@ def compare_records(
         pipeline.enqueue_sql(sql, "__splink__found_by_blocking_rules")
 
     predictions = db_api.sql_pipeline_to_splink_dataframe(pipeline)
-    _sql_cache.set(settings, predictions.sql_used_to_create, uid)
+    _sql_cache.set(
+        settings, predictions.sql_used_to_create, uid, sql_dialect_str=sql_dialect_str
+    )
 
     return predictions

--- a/splink/internals/realtime.py
+++ b/splink/internals/realtime.py
@@ -1,7 +1,6 @@
 from __future__ import annotations
 
 import json
-import logging
 from pathlib import Path
 from typing import Any
 from weakref import ref
@@ -15,7 +14,6 @@ from splink.internals.predict import (
 )
 from splink.internals.settings_creator import SettingsCreator
 from splink.internals.splink_dataframe import SplinkDataFrame
-
 
 
 class SQLCache:
@@ -40,7 +38,6 @@ class SQLCache:
             del self._cache[settings_id]
             return None
 
-        sql, cached_uid = self._cache[settings]
         if cached_uid:
             sql = sql.replace(cached_uid, new_uid)
         return sql
@@ -138,9 +135,7 @@ def compare_records(
     df_records_right.templated_name = "__splink__compare_records_right"
 
     if use_sql_from_cache:
-        cached_sql = _sql_cache.get(
-            settings, uid, sql_dialect_str=sql_dialect_str
-        )
+        cached_sql = _sql_cache.get(settings, uid, sql_dialect_str=sql_dialect_str)
         if cached_sql:
             return db_api._sql_to_splink_dataframe(
                 cached_sql,

--- a/splink/internals/realtime.py
+++ b/splink/internals/realtime.py
@@ -2,6 +2,7 @@ from __future__ import annotations
 
 from pathlib import Path
 from typing import Any
+from weakref import WeakKeyDictionary
 
 from splink.internals.accuracy import _select_found_by_blocking_rules
 from splink.internals.database_api import AcceptableInputTableType, DatabaseAPISubClass
@@ -16,20 +17,25 @@ from splink.internals.splink_dataframe import SplinkDataFrame
 
 class SQLCache:
     def __init__(self):
-        self._cache: dict[int, tuple[str, str | None]] = {}
+        self._cache: WeakKeyDictionary = WeakKeyDictionary()
 
-    def get(self, settings_id: int, new_uid: str) -> str | None:
-        if settings_id not in self._cache:
+    # TODO: if we have path/string, do we want to think about behaviour if underlying
+    # file changes between calls?
+    # TODO: might need to think about how equality works with WeakKeyDictionary in case
+    # there are any gotchas
+    # TODO: maybe check if string interning affects this in some way
+    def get(self, settings: SettingsCreator | dict[str, Any] | Path | str, new_uid: str) -> str | None:
+        if settings not in self._cache:
             return None
 
-        sql, cached_uid = self._cache[settings_id]
+        sql, cached_uid = self._cache[settings]
         if cached_uid:
             sql = sql.replace(cached_uid, new_uid)
         return sql
 
-    def set(self, settings_id: int, sql: str | None, uid: str | None) -> None:
+    def set(self, settings: SettingsCreator | dict[str, Any] | Path | str, sql: str | None, uid: str | None) -> None:
         if sql is not None:
-            self._cache[settings_id] = (sql, uid)
+            self._cache[settings] = (sql, uid)
 
 
 _sql_cache = SQLCache()
@@ -84,7 +90,7 @@ def compare_records(
 
     settings_id = id(settings)
     if use_sql_from_cache:
-        if cached_sql := _sql_cache.get(settings_id, uid):
+        if cached_sql := _sql_cache.get(settings, uid):
             return db_api._sql_to_splink_dataframe(
                 cached_sql,
                 templated_name="__splink__realtime_compare_records",

--- a/splink/internals/realtime.py
+++ b/splink/internals/realtime.py
@@ -2,7 +2,7 @@ from __future__ import annotations
 
 from pathlib import Path
 from typing import Any
-from weakref import WeakKeyDictionary
+from weakref import ref
 
 from splink.internals.accuracy import _select_found_by_blocking_rules
 from splink.internals.database_api import AcceptableInputTableType, DatabaseAPISubClass
@@ -17,7 +17,7 @@ from splink.internals.splink_dataframe import SplinkDataFrame
 
 class SQLCache:
     def __init__(self):
-        self._cache: WeakKeyDictionary = WeakKeyDictionary()
+        self._cache = {}
 
     # TODO: if we have path/string, do we want to think about behaviour if underlying
     # file changes between calls?
@@ -25,7 +25,13 @@ class SQLCache:
     # there are any gotchas
     # TODO: maybe check if string interning affects this in some way
     def get(self, settings: SettingsCreator | dict[str, Any] | Path | str, new_uid: str) -> str | None:
-        if settings not in self._cache:
+        settings_id = id(settings)
+        if settings_id not in self._cache:
+            return None
+        sql, cached_uid, settings_ref = self._cache[settings_id]
+        # if reference is dead, delete cache entry and return nowt
+        if settings_ref() is None:
+            del self._cache[settings_id]
             return None
 
         sql, cached_uid = self._cache[settings]
@@ -34,8 +40,9 @@ class SQLCache:
         return sql
 
     def set(self, settings: SettingsCreator | dict[str, Any] | Path | str, sql: str | None, uid: str | None) -> None:
+        settings_id = id(settings)
         if sql is not None:
-            self._cache[settings] = (sql, uid)
+            self._cache[settings_id] = (sql, uid, ref(settings))
 
 
 _sql_cache = SQLCache()
@@ -143,6 +150,6 @@ def compare_records(
         pipeline.enqueue_sql(sql, "__splink__found_by_blocking_rules")
 
     predictions = db_api.sql_pipeline_to_splink_dataframe(pipeline)
-    _sql_cache.set(settings_id, predictions.sql_used_to_create, uid)
+    _sql_cache.set(settings, predictions.sql_used_to_create, uid)
 
     return predictions

--- a/tests/test_realtime.py
+++ b/tests/test_realtime.py
@@ -354,3 +354,70 @@ def test_realtime_cache_different_settings(test_helpers, dialect):
         df1, df2, settings_1, db_api, use_sql_from_cache=True
     ).as_record_dict()[0]["match_weight"]
     assert res1 == pytest.approx(res1_again)
+
+
+@mark_with_dialects_excluding()
+def test_realtime_cache_different_settings_dict(test_helpers, dialect):
+    helper = test_helpers[dialect]
+    db_api = helper.extra_linker_args()["db_api"]
+
+    df1 = pd.DataFrame(
+        [
+            {
+                "unique_id": 0,
+                "first_name": "Julia",
+                "surname": "Taylor",
+                "city": "London",
+                "email": "julia@email.com",
+            }
+        ]
+    )
+
+    df2 = pd.DataFrame(
+        [
+            {
+                "unique_id": 1,
+                "first_name": "Julia",
+                "surname": "Taylor",
+                "city": "London",
+                "email": "bad@address.com",
+            }
+        ]
+    )
+
+    settings_1 = {
+        "link_type": "dedupe_only",
+        "comparisons": [
+            cl.ExactMatch("first_name"),
+            cl.ExactMatch("surname"),
+            cl.ExactMatch("city"),
+        ],
+        "blocking_rules_to_generate_predictions": [block_on("first_name")],
+    }
+
+    settings_2 = {
+        "link_type": "dedupe_only",
+        "comparisons": [
+            cl.ExactMatch("first_name"),
+            cl.ExactMatch("surname"),
+            cl.ExactMatch("email"),
+        ],
+        "blocking_rules_to_generate_predictions": [block_on("first_name")],
+    }
+
+    res1 = compare_records(df1, df2, settings_1, db_api, use_sql_from_cache=True)
+    res1 = res1.as_record_dict()[0]["match_weight"]
+
+    res2 = compare_records(df1, df2, settings_2, db_api, use_sql_from_cache=True)
+    res2 = res2.as_record_dict()[0]["match_weight"]
+
+    # should be different results as different model
+    assert res1 != pytest.approx(res2)
+
+    res1_again = compare_records(
+        df1, df2, settings_1, db_api, use_sql_from_cache=True
+    )
+    res1_again = res1_again.as_record_dict()[0]["match_weight"]
+    # using cache
+    assert res1 == pytest.approx(res1_again)
+

--- a/tests/test_realtime.py
+++ b/tests/test_realtime.py
@@ -414,10 +414,7 @@ def test_realtime_cache_different_settings_dict(test_helpers, dialect):
     # should be different results as different model
     assert res1 != pytest.approx(res2)
 
-    res1_again = compare_records(
-        df1, df2, settings_1, db_api, use_sql_from_cache=True
-    )
+    res1_again = compare_records(df1, df2, settings_1, db_api, use_sql_from_cache=True)
     res1_again = res1_again.as_record_dict()[0]["match_weight"]
     # using cache
     assert res1 == pytest.approx(res1_again)
-

